src/types/text.js

```diff
@@ -0,0 +1,189 @@
+    const {x, y, width, height} = this.getProps(['x', 'y', 'width', 'height']);
+
+    return this.labelIsVisible() && mouseX >= x &&
+			mouseX <= x + width &&
+			mouseY >= y &&
+			mouseY <= y + height;
+  }
+
+  labelIsVisible() {
+    return this.options.display && this.options.content;
+  }
+
+  boxMustBeDrawn() {
+    return (this.options.backgroundColor && this.options.backgroundColor !== 'transparent') || this.options.borderWidth > 0;
+  }
+
+  getCenterPoint(useFinalPosition) {
+    return getCenterPoint(this, useFinalPosition);
+  }
+
+  draw(ctx) {
+    if (this.labelIsVisible()) {
+      if (this.boxMustBeDrawn()) {
+        drawBox(ctx, this);
+      }
+      ctx.save();
+      drawLabel(ctx, this);
+      ctx.restore();
+    }
+  }
+
+  resolveElementProperties(chart, options) {
+    const {chartArea, scales, ctx} = chart;
+    const xScale = scales[options.xScaleID];
+    const yScale = scales[options.yScaleID];
+    let x = chartArea.width / 2;
+    let y = chartArea.height / 2;
+
+    if (xScale) {
+      x = scaleValue(xScale, options.xValue, x);
+    }
+
+    if (yScale) {
+      y = scaleValue(yScale, options.yValue, y);
+    }
+    return calculateRect(x, y, measureLabel(ctx, options), options);
+  }
+}
+
+TextAnnotation.id = 'textAnnotation';
+
+TextAnnotation.defaults = {
+  adjustScaleRange: true,
+  align: 'center',
+  display: true,
+  backgroundColor: 'transparent',
+  borderDash: [],
+  borderDashOffset: 0,
+  borderWidth: 0,
+  borderRadius: 0,
+  color: 'black',
+  content: null,
+  drawTime: undefined,
+  font: {
+    family: undefined,
+    lineHeight: undefined,
+    size: undefined,
+    style: undefined,
+    weight: undefined
+  },
+  position: 'middle',
+  textAlign: 'center',
+  xAdjust: 0,
+  xPadding: 6,
+  xScaleID: 'x',
+  xValue: undefined,
+  yAdjust: 0,
+  yPadding: 6,
+  yScaleID: 'y',
+  yValue: undefined
+};
+
+TextAnnotation.defaultRoutes = {
+  borderColor: 'color',
+};
+
+function drawBox(ctx, text) {
+  const {x, y, width, height, options} = text.getProps(['x', 'y', 'width', 'height', 'options']);
+  ctx.save();
+  ctx.lineWidth = options.borderWidth;
+  ctx.strokeStyle = options.borderColor;
+  ctx.fillStyle = options.backgroundColor;
+  ctx.setLineDash(options.borderDash);
+  ctx.lineDashOffset = options.borderDashOffset;
+  ctx.beginPath();
+  addRoundedRectPath(ctx, {
+    x, y, w: width, h: height,
+    radius: clampAll(toTRBLCorners(options.borderRadius), 0, Math.min(width, height) / 2)
+  });
+  ctx.closePath();
+  ctx.fill();
+  if (options.borderWidth) {
+    ctx.stroke();
+  }
+  ctx.restore();
+}
+
+const widthCache = new Map();
+function measureLabel(ctx, label) {
+  const content = label.content;
+  const lines = isArray(content) ? content : [content];
+  const count = lines.length;
+  const font = toFont(label.font);
+  const lh = font.lineHeight;
+  const xPadding = label.xPadding;
+  const yPadding = label.yPadding;
+  const borderWidth = label.borderWidth;
```

Comment:
These are only used once, so no need to cache in variables.

Reply:
done!